File: ios/SCMessenger/SCMessenger/Transport/SmartTransportRouter.swift
//
//  SmartTransportRouter.swift
//  SCMessenger
//
//  Smart transport selection with 500ms timeout fallback and transport health tracking.
//  Implements parallel transport racing for optimal message delivery latency.
//

import Foundation
import os

/// Result of a transport delivery attempt
struct TransportDeliveryResult {
    let transport: TransportType
    let success: Bool
    let latencyMs: UInt64
    let error: String?
    let timestamp: Date
}

/// Health metrics for a specific transport to a specific peer
struct TransportHealth: Codable {
    var lastSuccessAt: Date?
    var lastFailureAt: Date?
    var successCount: UInt64 = 0
    var failureCount: UInt64 = 0
    var averageLatencyMs: Double = 0
    var lastLatencyMs: UInt64 = 0
    
    /// Success rate (0.0 to 1.0)
    var successRate: Double {
        let total = successCount + failureCount
        guard total > 0 else { return 0.5 } // Unknown = neutral
        return Double(successCount) / Double(total)
    }
    
    /// Is this transport considered healthy?
    var isHealthy: Bool {
        // Healthy if success rate > 50% and not too many recent failures
        guard successRate > 0.5 else { return false }
        // If last failure was very recent (< 5s), be cautious
        if let lastFailure = lastFailureAt {
            return Date().timeIntervalSince(lastFailure) > 5.0
        }
        return true
    }
    
    /// Score for transport selection (higher = better)
    var score: Double {
        // Weight: 70% success rate, 30% latency (inverted, lower latency = higher score)
        let latencyScore = averageLatencyMs > 0 ? min(1.0, 500.0 / averageLatencyMs) : 0.5
        return (successRate * 0.7) + (latencyScore * 0.3)
    }
}

/// Message deduplication entry
struct MessageDedupEntry {
    let messageId: String
    let firstReceivedAt: Date
    let firstTransport: TransportType
    var duplicateCount: UInt32 = 0
    var duplicateTimestamps: [Date] = []
    var duplicateTransports: [TransportType] = []
}

/// Smart transport router with health tracking and parallel racing
@MainActor
final class SmartTransportRouter {
    private let logger = Logger(subsystem: "com.scmessenger", category: "TransportRouter")
    
    // Transport health tracking per peer
    private var transportHealth: [String: [TransportType: TransportHealth]] = [:]
    
    // Message deduplication cache
    private var messageDedupCache: [String: MessageDedupEntry] = [:]
    private let dedupCacheTtl: TimeInterval = 300 // 5 minutes
    
    // Timeout for "preferred" transport before racing all
    private let preferredTransportTimeoutMs: UInt64 = 500
    
    // Last successful transport per peer (for "previously used/good path")
    private var lastSuccessfulTransport: [String: TransportType] = [:]
    
    // MARK: - Transport Health Management
    
    /// Record a successful delivery
    func recordSuccess(peerId: String, transport: TransportType, latencyMs: UInt64) {
        var health = getHealth(peerId: peerId, transport: transport)
        health.lastSuccessAt = Date()
        health.successCount += 1
        health.lastLatencyMs = latencyMs
        
        // Update rolling average latency
        let totalDeliveries = health.successCount + health.failureCount
        if totalDeliveries > 1 {
            health.averageLatencyMs = ((health.averageLatencyMs * Double(totalDeliveries - 1)) + Double(latencyMs)) / Double(totalDeliveries)
        } else {
            health.averageLatencyMs = Double(latencyMs)
        }
        
        setHealth(peerId: peerId, transport: transport, health: health)
        lastSuccessfulTransport[peerId] = transport
        
        logger.info("Transport health updated: peer=\(peerId.prefix(8)) transport=\(transport.rawValue) success rate=\(String(format: "%.2f", health.successRate)) avg_latency=\(String(format: "%.0f", health.averageLatencyMs))ms")
    }
    
    /// Record a failed delivery
    func recordFailure(peerId: String, transport: TransportType, error: String?) {
        var health = getHealth(peerId: peerId, transport: transport)
        health.lastFailureAt = Date()
        health.failureCount += 1
        
        setHealth(peerId: peerId, transport: transport, health: health)
        
        logger.warning("Transport failure: peer=\(peerId.prefix(8)) transport=\(transport.rawValue) error=\(error ?? "unknown") success rate=\(String(format: "%.2f", health.successRate))")
    }
    
    /// Get health for a specific peer and transport
    private func getHealth(peerId: String, transport: TransportType) -> TransportHealth {
        return transportHealth[peerId]?[transport] ?? TransportHealth()
    }
    
    /// Set health for a specific peer and transport
    private func setHealth(peerId: String, transport: TransportType, health: TransportHealth) {
        if transportHealth[peerId] == nil {
            transportHealth[peerId] = [:]
        }
        transportHealth[peerId]?[transport] = health
    }
    
    // MARK: - Transport Selection
    
    /// Get the preferred transport for a peer (previously successful or highest score)
    func getPreferredTransport(peerId: String) -> TransportType? {
        // First check if we have a last successful transport
        if let lastSuccess = lastSuccessfulTransport[peerId] {
            let health = getHealth(peerId: peerId, transport: lastSuccess)
            if health.isHealthy {
                return lastSuccess
            }
        }
        
        // Otherwise, find the transport with the highest score
        let allTransports = TransportType.allCases
        var bestTransport: TransportType?
        var bestScore: Double = -1
        
        for transport in allTransports {
            let health = getHealth(peerId: peerId, transport: transport)
            if health.score > bestScore {
                bestScore = health.score
                bestTransport = transport
            }
        }
        
        return bestTransport
    }
    
    /// Get all available transports sorted by score (best first)
    func getAvailableTransportsSorted(peerId: String) -> [TransportType] {
        return TransportType.allCases.sorted { transport1, transport2 in
            let health1 = getHealth(peerId: peerId, transport: transport1)
            let health2 = getHealth(peerId: peerId, transport: transport2)
            return health1.score > health2.score
        }
    }
    
    // MARK: - Message Deduplication
    
    /// Check if a message is a duplicate and record it
    /// Returns: (isDuplicate, timeVarianceMs, firstTransport)
    func checkAndRecordMessage(
        messageId: String,
        transport: TransportType
    ) -> (isDuplicate: Bool, timeVarianceMs: UInt64?, firstTransport: TransportType?) {
        let now = Date()
        
        // Clean up old entries
        cleanupDedupCache()
        
        if let existing = messageDedupCache[messageId] {
            // This is a duplicate
            let timeVarianceMs = UInt64(now.timeIntervalSince(existing.firstReceivedAt) * 1000)
            
            var updated = existing
            updated.duplicateCount += 1
            updated.duplicateTimestamps.append(now)
            updated.duplicateTransports.append(transport)
            messageDedupCache[messageId] = updated
            
            logger.info("Message duplicate detected: msg=\(messageId.prefix(8)) transport=\(transport.rawValue) time_variance=\(timeVarianceMs)ms first_transport=\(existing.firstTransport.rawValue) duplicate_count=\(updated.duplicateCount)")
            
            return (isDuplicate: true, timeVarianceMs: timeVarianceMs, firstTransport: existing.firstTransport)
        } else {
            // First receipt of this message
            let entry = MessageDedupEntry(
                messageId: messageId,
                firstReceivedAt: now,
                firstTransport: transport,
                duplicateCount: 0,
                duplicateTimestamps: [],
                duplicateTransports: []
            )
            messageDedupCache[messageId] = entry
            
            logger.info("Message first receipt: msg=\(messageId.prefix(8)) transport=\(transport.rawValue)")
            
            return (isDuplicate: false, timeVarianceMs: nil, firstTransport: nil)
        }
    }
    
    /// Get dedup statistics for a message (for mesh enhancement logging)
    func getDedupStats(messageId: String) -> MessageDedupEntry? {
        return messageDedupCache[messageId]
    }
    
    /// Clean up old dedup cache entries
    private func cleanupDedupCache() {
        let cutoff = Date().addingTimeInterval(-dedupCacheTtl)
        messageDedupCache = messageDedupCache.filter { _, entry in
            entry.firstReceivedAt > cutoff
        }
    }
    
    // MARK: - Smart Delivery
    
    /// Attempt delivery with smart transport selection
    /// - Tries preferred transport first
    /// - If no response within 500ms, races all available transports
    /// - Returns the first successful result
    func attemptDelivery(
        peerId: String,
        envelopeData: Data,
        multipeerPeerId: String?,
        blePeerId: String?,
        tcpMdnsPeerId: String?,
        routePeerCandidates: [String],
        addresses: [String],
        traceMessageId: String?,
        attemptContext: String?,
        tryMultipeer: @escaping (String) async -> Bool,
        tryBle: @escaping (String) async -> Bool,
        tryTcpMdns: @escaping (String) async -> Bool,
        tryCore: @escaping (String) async -> Bool
    ) async -> TransportDeliveryResult {
        let startTime = Date()
        
        // Determine available transports
        var availableTransports: [(type: TransportType, target: String, attempt: () async -> Bool)] = []
        
        if let multipeerTarget = multipeerPeerId?.trimmingCharacters(in: .whitespacesAndNewlines), !multipeerTarget.isEmpty {
            availableTransports.append((.multipeer, multipeerTarget, { await tryMultipeer(multipeerTarget) }))
        }
        
        if let bleTarget = blePeerId?.trimmingCharacters(in: .whitespacesAndNewlines), !bleTarget.isEmpty {
            availableTransports.append((.ble, bleTarget, { await tryBle(bleTarget) }))
        }
        
        if let tcpMdnsTarget = tcpMdnsPeerId?.trimmingCharacters(in: .whitespacesAndNewlines), !tcpMdnsTarget.isEmpty {
            availableTransports.append((.tcpMdns, tcpMdnsTarget, { await tryTcpMdns(tcpMdnsTarget) }))
        }
        
        if let internetTarget = routePeerCandidates.first?.trimmingCharacters(in: .whitespacesAndNewlines), !internetTarget.isEmpty {
            availableTransports.append((.internet, internetTarget, { await tryCore(internetTarget) }))
        }
        
        guard !availableTransports.isEmpty else {
            logger.warning("No available transports for peer \(peerId.prefix(8))")
            return TransportDeliveryResult(
                transport: .internet,
                success: false,
                latencyMs: 0,
                error: "no_available_transports",
                timestamp: Date()
            )
        }
        
        // Get preferred transport
        let preferredTransport = getPreferredTransport(peerId: peerId)
        
        // If we have a preferred transport, try it first with timeout
        if let preferred = preferredTransport,
           let preferredAttempt = availableTransports.first(where: { $0.type == preferred }) {
            logger.info("Trying preferred transport \(preferred.rawValue) for peer \(peerId.prefix(8))")
            
            // Race preferred transport against timeout
            let preferredResult = await withTaskGroup(of: Bool?.self, returning: Bool?.self) { group in
                // Preferred transport attempt
                group.addTask {
                    await preferredAttempt.attempt()
                }
                
                // Timeout task
                group.addTask {
                    try? await Task.sleep(nanoseconds: self.preferredTransportTimeoutMs * 1_000_000)
                    return nil // Timeout signal
                }
                
                // Wait for first result
                let result = await group.next()
                group.cancelAll()
                return result ?? nil
            }
            
            if let success = preferredResult, success {
                let latencyMs = UInt64(Date().timeIntervalSince(startTime) * 1000)
                recordSuccess(peerId: peerId, transport: preferred, latencyMs: latencyMs)
                logger.info("✓ Preferred transport \(preferred.rawValue) succeeded in \(latencyMs)ms")
                return TransportDeliveryResult(
                    transport: preferred,
                    success: true,
                    latencyMs: latencyMs,
                    error: nil,
                    timestamp: Date()
                )
            }
            
            // Preferred transport failed or timed out - race all transports
            logger.warning("Preferred transport \(preferred.rawValue) failed/timed out, racing all transports")
        }
        
        // Race all available transports in parallel
        logger.info("Racing \(availableTransports.count) transports for peer \(peerId.prefix(8))")
        
        let result = await withTaskGroup(
            of: (transport: TransportType, success: Bool, latencyMs: UInt64)?.self,
            returning: (transport: TransportType, success: Bool, latencyMs: UInt64)?.self
        ) { group in
            for transportAttempt in availableTransports {
                group.addTask {
                    let transportStart = Date()
                    let success = await transportAttempt.attempt()
                    let latencyMs = UInt64(Date().timeIntervalSince(transportStart) * 1000)
                    return (transport: transportAttempt.type, success: success, latencyMs: latencyMs)
                }
            }
            
            // Return first successful result
            for await result in group {
                if let result = result, result.success {
                    group.cancelAll()
                    return result
                }
            }
            
            return nil
        }
        
        if let result = result {
            recordSuccess(peerId: peerId, transport: result.transport, latencyMs: result.latencyMs)
            logger.info("✓ Transport \(result.transport.rawValue) succeeded in \(result.latencyMs)ms")
            return TransportDeliveryResult(
                transport: result.transport,
                success: true,
                latencyMs: result.latencyMs,
                error: nil,
                timestamp: Date()
            )
        } else {
            // All transports failed
            let latencyMs = UInt64(Date().timeIntervalSince(startTime) * 1000)
            for transportAttempt in availableTransports {
                recordFailure(peerId: peerId, transport: transportAttempt.type, error: "all_transports_failed")
            }
            logger.error("✗ All transports failed for peer \(peerId.prefix(8))")
            return TransportDeliveryResult(
                transport: .internet,
                success: false,
                latencyMs: latencyMs,
                error: "all_transports_failed",
                timestamp: Date()
            )
        }
    }
    
    // MARK: - Diagnostics
    
    /// Get transport health summary for diagnostics
    func getHealthSummary() -> [String: [String: Any]] {
        var summary: [String: [String: Any]] = [:]
        
        for (peerId, transports) in transportHealth {
            var peerSummary: [String: Any] = [:]
            for (transport, health) in transports {
                peerSummary[transport.rawValue] = [
                    "success_rate": health.successRate,
                    "avg_latency_ms": health.averageLatencyMs,
                    "success_count": health.successCount,
                    "failure_count": health.failureCount,
                    "is_healthy": health.isHealthy,
                    "score": health.score
                ]
            }
            summary[peerId] = peerSummary
        }
        
        return summary
    }
    
    /// Reset health for a peer (e.g., after reconnection)
    func resetHealth(peerId: String) {
        transportHealth.removeValue(forKey: peerId)
        lastSuccessfulTransport.removeValue(forKey: peerId)
        logger.info("Transport health reset for peer \(peerId.prefix(8))")
    }
}

File: android/app/src/main/java/com/scmessenger/android/transport/SmartTransportRouter.kt
package com.scmessenger.android.transport

import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import timber.log.Timber
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong

/**
 * Smart transport selection with 500ms timeout fallback and transport health tracking.
 * Implements parallel transport racing for optimal message delivery latency.
 * 
 * Mirrors: iOS/SCMessenger/SCMessenger/Transport/SmartTransportRouter.swift
 */
class SmartTransportRouter {

    companion object {
        private const val TAG = "SmartTransportRouter"
        private const val PREFERRED_TRANSPORT_TIMEOUT_MS = 500L
        private const val DEDUP_CACHE_TTL_MS = 300_000L // 5 minutes
    }

    /**
     * Transport types available for message delivery
     */
    enum class TransportType(val value: String) {
        WIFI_DIRECT("wifi_direct"),
        BLE("ble"),
        CORE("core"), // libp2p/internet relay
        TCP_MDNS("tcp_mdns"); // LAN-direct via libp2p TCP (mDNS-discovered peers)

        companion object {
            fun fromValue(value: String): TransportType? {
                return entries.find { it.value == value }
            }
        }
    }

    /**
     * Result of a transport delivery attempt
     */
    data class TransportDeliveryResult(
        val transport: TransportType,
        val success: Boolean,
        val latencyMs: Long,
        val error: String?,
        val timestamp: Long = System.currentTimeMillis()
    )

    /**
     * Health metrics for a specific transport to a specific peer
     */
    data class TransportHealth(
        var lastSuccessAt: Long? = null,
        var lastFailureAt: Long? = null,
        var successCount: Long = 0,
        var failureCount: Long = 0,
        var averageLatencyMs: Double = 0.0,
        var lastLatencyMs: Long = 0
    ) {
        /** Success rate (0.0 to 1.0) */
        val successRate: Double
            get() {
                val total = successCount + failureCount
                return if (total > 0) successCount.toDouble() / total.toDouble() else 0.5
            }

        /** Is this transport considered healthy? */
        val isHealthy: Boolean
            get() {
                if (successRate <= 0.5) return false
                // If last failure was very recent (< 5s), be cautious
                val lastFailure = lastFailureAt ?: return true
                return (System.currentTimeMillis() - lastFailure) > 5000
            }

        /** Score for transport selection (higher = better) */
        val score: Double
            get() {
                // Weight: 70% success rate, 30% latency (inverted, lower latency = higher score)
                val latencyScore = if (averageLatencyMs > 0) minOf(1.0, 500.0 / averageLatencyMs) else 0.5
                return (successRate * 0.7) + (latencyScore * 0.3)
            }
    }

    /**
     * Message deduplication entry
     */
    data class MessageDedupEntry(
        val messageId: String,
        val firstReceivedAt: Long,
        val firstTransport: TransportType,
        var duplicateCount: Int = 0,
        val duplicateTimestamps: MutableList<Long> = mutableListOf(),
        val duplicateTransports: MutableList<TransportType> = mutableListOf()
    )

    // Transport health tracking per peer
    private val transportHealth = ConcurrentHashMap<String, ConcurrentHashMap<TransportType, TransportHealth>>()
    
    // Message deduplication cache
    private val messageDedupCache = ConcurrentHashMap<String, MessageDedupEntry>()
    private val dedupMutex = Mutex()
    
    // Last successful transport per peer (for "previously used/good path")
    private val lastSuccessfulTransport = ConcurrentHashMap<String, TransportType>()
    
    // Mutex for health updates
    private val healthMutex = Mutex()

    // MARK: - Transport Health Management

    /**
     * Record a successful delivery
     */
    suspend fun recordSuccess(peerId: String, transport: TransportType, latencyMs: Long) {
        healthMutex.withLock {
            val health = getHealth(peerId, transport)
            health.lastSuccessAt = System.currentTimeMillis()
            health.successCount++
            health.lastLatencyMs = latencyMs

            // Update rolling average latency
            val totalDeliveries = health.successCount + health.failureCount
            health.averageLatencyMs = if (totalDeliveries > 1) {
                ((health.averageLatencyMs * (totalDeliveries - 1)) + latencyMs) / totalDeliveries
            } else {
                latencyMs.toDouble()
            }

            setHealth(peerId, transport, health)
            lastSuccessfulTransport[peerId] = transport

            Timber.tag(TAG).i("Transport health updated: peer=${peerId.take(8)} transport=${transport.value} success_rate=${String.format("%.2f", health.successRate)} avg_latency=${String.format("%.0f", health.averageLatencyMs)}ms")
        }
    }

    /**
     * Record a failed delivery
     */
    suspend fun recordFailure(peerId: String, transport: TransportType, error: String?) {
        healthMutex.withLock {
            val health = getHealth(peerId, transport)
            health.lastFailureAt = System.currentTimeMillis()
            health.failureCount++

            setHealth(peerId, transport, health)

            Timber.tag(TAG).w("Transport failure: peer=${peerId.take(8)} transport=${transport.value} error=${error ?: "unknown"} success_rate=${String.format("%.2f", health.successRate)}")
        }
    }

    /**
     * Get health for a specific peer and transport
     */
    private fun getHealth(peerId: String, transport: TransportType): TransportHealth {
        return transportHealth.getOrPut(peerId) { ConcurrentHashMap() }
            .getOrPut(transport) { TransportHealth() }
    }

    /**
     * Set health for a specific peer and transport
     */
    private fun setHealth(peerId: String, transport: TransportType, health: TransportHealth) {
        transportHealth.getOrPut(peerId) { ConcurrentHashMap() }[transport] = health
    }

    // MARK: - Transport Selection

    /**
     * Get the preferred transport for a peer (previously successful or highest score)
     */
    fun getPreferredTransport(peerId: String): TransportType? {
        // First check if we have a last successful transport
        val lastSuccess = lastSuccessfulTransport[peerId]
        if (lastSuccess != null) {
            val health = getHealth(peerId, lastSuccess)
            if (health.isHealthy) {
                return lastSuccess
            }
        }

        // Otherwise, find the transport with the highest score
        var bestTransport: TransportType? = null
        var bestScore = -1.0

        for (transport in TransportType.entries) {
            val health = getHealth(peerId, transport)
            if (health.score > bestScore) {
                bestScore = health.score
                bestTransport = transport
            }
        }

        return bestTransport
    }

    /**
     * Get all available transports sorted by score (best first)
     */
    fun getAvailableTransportsSorted(peerId: String): List<TransportType> {
        return TransportType.entries.sortedByDescending { transport ->
            getHealth(peerId, transport).score
        }
    }

    /**
     * Get the list of available transport types for a peer.
     * Returns all transport types that have recorded health data or
     * that are potentially available (no failures recorded).
     */
    fun getAvailableTransports(peerId: String): List<TransportType> {
        val peerTransports = transportHealth[peerId]
        return if (peerTransports != null && peerTransports.isNotEmpty()) {
            // Return transports that have health data (indicating they've been used)
            peerTransports.keys.toList()
        } else {
            // No health data yet — return all transport types as potentially available
            TransportType.entries
        }
    }

    // MARK: - Message Deduplication

    /**
     * Check if a message is a duplicate and record it
     * @return Triple(isDuplicate, timeVarianceMs, firstTransport)
     */
    suspend fun checkAndRecordMessage(
        messageId: String,
        transport: TransportType
    ): Triple<Boolean, Long?, TransportType?> {
        return dedupMutex.withLock {
            val now = System.currentTimeMillis()

            // Clean up old entries
            cleanupDedupCache()

            val existing = messageDedupCache[messageId]
            if (existing != null) {
                // This is a duplicate
                val timeVarianceMs = now - existing.firstReceivedAt

                existing.duplicateCount++
                existing.duplicateTimestamps.add(now)
                existing.duplicateTransports.add(transport)

                Timber.tag(TAG).i("Message duplicate detected: msg=${messageId.take(8)} transport=${transport.value} time_variance=${timeVarianceMs}ms first_transport=${existing.firstTransport.value} duplicate_count=${existing.duplicateCount}")

                Triple(true, timeVarianceMs, existing.firstTransport)
            } else {
                // First receipt of this message
                val entry = MessageDedupEntry(
                    messageId = messageId,
                    firstReceivedAt = now,
                    firstTransport = transport,
                    duplicateCount = 0,
                    duplicateTimestamps = mutableListOf(),
                    duplicateTransports = mutableListOf()
                )
                messageDedupCache[messageId] = entry

                Timber.tag(TAG).i("Message first receipt: msg=${messageId.take(8)} transport=${transport.value}")

                Triple(false, null, null)
            }
        }
    }

    /**
     * Get dedup statistics for a message (for mesh enhancement logging)
     */
    fun getDedupStats(messageId: String): MessageDedupEntry? {
        return messageDedupCache[messageId]
    }

    /**
     * Clean up old dedup cache entries
     */
    private fun cleanupDedupCache() {
        val cutoff = System.currentTimeMillis() - DEDUP_CACHE_TTL_MS
        messageDedupCache.entries.removeIf { (_, entry) ->
            entry.firstReceivedAt < cutoff
        }
    }

    // MARK: - Smart Delivery

    /**
     * Attempt delivery with smart transport selection
     * - Tries preferred transport first
     * - If no response within 500ms, races all available transports
     * - Returns the first successful result
     */
    suspend fun attemptDelivery(
        peerId: String,
        @Suppress("UNUSED_PARAMETER") envelopeData: ByteArray,
        wifiPeerId: String?,
        blePeerId: String?,
        tcpMdnsPeerId: String?,
        routePeerCandidates: List<String>,
        @Suppress("UNUSED_PARAMETER") listeners: List<String>,
        @Suppress("UNUSED_PARAMETER") traceMessageId: String?,
        @Suppress("UNUSED_PARAMETER") attemptContext: String?,
        tryWifi: suspend (String) -> Boolean,
        tryBle: suspend (String) -> Boolean,
        tryTcpMdns: suspend (String) -> Boolean,
        tryCore: suspend (String) -> Boolean
    ): TransportDeliveryResult {
        val startTime = System.currentTimeMillis()

        // Determine available transports
        data class TransportAttempt(
            val type: TransportType,
            val target: String,
            val attempt: suspend () -> Boolean
        )

        val availableTransports = mutableListOf<TransportAttempt>()

        val wifiTarget = wifiPeerId?.trim()?.takeIf { it.isNotEmpty() }
        if (wifiTarget != null) {
            availableTransports.add(TransportAttempt(TransportType.WIFI_DIRECT, wifiTarget) { tryWifi(wifiTarget) })
        }

        val bleTarget = blePeerId?.trim()?.takeIf { it.isNotEmpty() }
        if (bleTarget != null) {
            availableTransports.add(TransportAttempt(TransportType.BLE, bleTarget) { tryBle(bleTarget) })
        }

        val tcpMdnsTarget = tcpMdnsPeerId?.trim()?.takeIf { it.isNotEmpty() }
        if (tcpMdnsTarget != null) {
            availableTransports.add(TransportAttempt(TransportType.TCP_MDNS, tcpMdnsTarget) { tryTcpMdns(tcpMdnsTarget) })
        }

        val coreTarget = routePeerCandidates.firstOrNull()?.trim()?.takeIf { it.isNotEmpty() }
        if (coreTarget != null) {
            availableTransports.add(TransportAttempt(TransportType.CORE, coreTarget) { tryCore(coreTarget) })
        }

        if (availableTransports.isEmpty()) {
            Timber.tag(TAG).w("No available transports for peer ${peerId.take(8)}")
            return TransportDeliveryResult(
                transport = TransportType.CORE,
                success = false,
                latencyMs = 0,
                error = "no_available_transports"
            )
        }

        // Get preferred transport
        val preferredTransport = getPreferredTransport(peerId)

        // If we have a preferred transport, try it first with timeout
        if (preferredTransport != null) {
            val preferredAttempt = availableTransports.find { it.type == preferredTransport }
            if (preferredAttempt != null) {
                Timber.tag(TAG).i("Trying preferred transport ${preferredTransport.value} for peer ${peerId.take(8)}")

                // Race preferred transport against timeout
                val preferredResult = withTimeoutOrNull(PREFERRED_TRANSPORT_TIMEOUT_MS) {
                    preferredAttempt.attempt()
                }

                if (preferredResult == true) {
                    val latencyMs = System.currentTimeMillis() - startTime
                    recordSuccess(peerId, preferredTransport, latencyMs)
                    Timber.tag(TAG).i("✓ Preferred transport ${preferredTransport.value} succeeded in ${latencyMs}ms")
                    return TransportDeliveryResult(
                        transport = preferredTransport,
                        success = true,
                        latencyMs = latencyMs,
                        error = null
                    )
                }

                // Preferred transport failed or timed out - race all transports
                Timber.tag(TAG).w("Preferred transport ${preferredTransport.value} failed/timed out, racing all transports")
            }
        }

        // Race all available transports in parallel
        Timber.tag(TAG).i("Racing ${availableTransports.count()} transports for peer ${peerId.take(8)}")

        val result = coroutineScope {
            val deferreds = availableTransports.map { transportAttempt ->
                async {
                    val transportStart = System.currentTimeMillis()
                    val success = try {
                        transportAttempt.attempt()
                    } catch (e: Exception) {
                        Timber.tag(TAG).w(e, "Transport ${transportAttempt.type.value} failed")
                        false
                    }
                    val latencyMs = System.currentTimeMillis() - transportStart
                    Triple(transportAttempt.type, success, latencyMs)
                }
            }

            // Wait for first successful result
            var firstSuccess: Triple<TransportType, Boolean, Long>? = null
            for (deferred in deferreds) {
                val result = deferred.await()
                if (result.second) {
                    firstSuccess = result
                    break
                }
            }

            // Cancel remaining coroutines
            deferreds.forEach { it.cancel() }

            firstSuccess
        }

        return if (result != null) {
            recordSuccess(peerId, result.first, result.third)
            Timber.tag(TAG).i("✓ Transport ${result.first.value} succeeded in ${result.third}ms")
            TransportDeliveryResult(
                transport = result.first,
                success = true,
                latencyMs = result.third,
                error = null
            )
        } else {
            // All transports failed
            val latencyMs = System.currentTimeMillis() - startTime
            availableTransports.forEach { transportAttempt ->
                recordFailure(peerId, transportAttempt.type, "all_transports_failed")
            }
            Timber.tag(TAG).e("✗ All transports failed for peer ${peerId.take(8)}")
            TransportDeliveryResult(
                transport = TransportType.CORE,
                success = false,
                latencyMs = latencyMs,
                error = "all_transports_failed"
            )
        }
    }

    // MARK: - Diagnostics

    /**
     * Get transport health summary for diagnostics
     */
    fun getHealthSummary(): Map<String, Map<String, Any>> {
        val summary = mutableMapOf<String, Map<String, Any>>()

        for ((peerId, transports) in transportHealth) {
            val peerSummary = mutableMapOf<String, Any>()
            for ((transport, health) in transports) {
                peerSummary[transport.value] = mapOf(
                    "success_rate" to health.successRate,
                    "avg_latency_ms" to health.averageLatencyMs,
                    "success_count" to health.successCount,
                    "failure_count" to health.failureCount,
                    "is_healthy" to health.isHealthy,
                    "score" to health.score
                )
            }
            summary[peerId] = peerSummary
        }

        return summary
    }

    /**
     * Reset health for a peer (e.g., after reconnection)
     */
    fun resetHealth(peerId: String) {
        transportHealth.remove(peerId)
        lastSuccessfulTransport.remove(peerId)
        Timber.tag(TAG).i("Transport health reset for peer ${peerId.take(8)}")
    }
}
